mautic/mautic
# Feature webhook authenticity verification

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | No |
| New feature? | Yes |
| Automated tests included? | No |
| Related user documentation PR URL |  No |
| Related developer documentation PR URL | https://github.com/mautic/developer-documentation/pull/114 |
| Issues addressed (#s or URLs) | Resolves #6734 | 
| BC breaks? | No |
| Deprecations? | No |

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
Add solution to verify webhook authenticity.
A secret is provided or generated on webhook creation. On each webhook call a base64 encoded HMAC-SHA256 signature is computed (with this secret) on the payload data. This signature is added in the header `Webhook-Signature`.
So a third party application (sharing the webhook secret) can compute the signature on the webhook body and verify the signature on the header to be sure that the webhook come from mautic and has not been modified.

For existing webhooks, the migration file automatically generate secret, so mautic can also generate signature for these webhooks.

#### Steps to test this PR:
1. Create a new webhook with `SECRET`
2. Start an api that implement webhook authenticity verification (example in nodeJS below)
3. Trigger the webhook
4. Verify api logs

#### NodeJS implementation example
```javascript
'use strict';

const express = require('express');
const crypto = require('crypto');
const app = express();
const port = 3000;
const SECRET = 'SECRET';

// save raw body
app.use ((req, res, next) => {
    let data = '';
    req.setEncoding('utf8');

    req.on('data', chunk => data += chunk);
    req.on('end', () => {
        req.body = data;
        return next();
    });
});

app.post('/webhook', (req, res) => {
    const receivedSignature = req.headers['webhook-signature'];
    console.log('Received signature (in header):', receivedSignature);

    const computedSignature = crypto.createHmac('sha256', SECRET).update(req.body).digest('base64');
    console.log('Computed signature (from body):', computedSignature);

    if (receivedSignature === computedSignature) {
        console.log('Webhook authenticity verification OK');
    } else {
        console.log('Webhook not authentic!');
    }

    res.send();
});

app.listen(port, () => console.log(`App listening on port ${port}!`));
```

#### List deprecations along with the new alternative:
No deprecations

#### List backwards compatibility breaks:
No backwards compatibility breaks
